33 updates to using git from rstudio (#49)
UCL-ARC/r-amr-epidemiology: episodes/04-git-changes.md

![](fig/RStudio_git_first_commit.png){alt='RStudio screenshot showing initial commit message for index.qmd.'}

To get it to do that, we need to type a message in the `Commit message` box and click on `Commit`.

On the completion of the `commit` the following message is displayed in the terminal:

```output
>>> /usr/bin/git commit -F /var/folders/1v/6jw6cbqn1sldcbfrc24vj5qm0000gn/T/RtmpRsFsk9/git-commit-message-617d1bbdcbbe.txt
[main (root-commit) a639e37] Added about information for the web page.
 1 file changed, 13 insertions(+)
 create mode 100644 index.qmd

```

When we run `commit`,
Git takes everything we have told it to save by using `stage`
and stores a copy permanently inside the special `.git` directory.

This permanent copy is called a [commit](../learners/reference.md#commit)
(or [revision](../learners/reference.md#revision)) and its short identifier is `a639e37`. Your commit may have another identifier.

We use the commit message to record a short, descriptive, and specific comment that will help us remember later on what we did and why.

[Good commit messages][commit-messages] start with a brief (\<50 characters) statement about the

> changes made in the commit. Generally, the message should complete the sentence "If applied, this commit will" <commit message here>.

Reviewer: ```suggestion
changes made in the commit. Generally, the message should complete the sentence "If applied, this commit will" `<commit message here>`.
```
Maybe setting it us code won't make the `<`/`>` characters to conflict?

If you want to go into more detail, add a blank line between the summary line and your additional notes. Use this additional space to explain why you made changes and/or what their impact will be.

We can select `Diff "index.qmd"` from the Git dropdowb menu to show us the current status of the file `index.qmd`.

![](fig/rs_git_diff_index_qmd.png){alt='RStudio screenshot showing selection of Diff index.qmd on Git dropdown menu.'}

As everything is up to date, a dialogue box appears informing us that there have been no changes to the file.

> ![](fig/rs_git_diff_no_change.png){alt='RStudio screenshot showing a dialogue box with the text "There are no changes to the file "index.qmd" to diff.".'}

Reviewer: ```suggestion
![](fig/rs_git_diff_no_change.png){alt='RStudio screenshot showing a dialogue box with the text "There are no changes to the file \"index.qmd\" to diff.".'}
```

I think... otherwise the first quotes will be finishing before index, and then start after the file name.

Git has a special *staging area*
where it keeps track of things that have been added to
the current [changeset](../learners/reference.md#changeset)
but not yet committed.

:::::::::::::::::::::::::::::::::::::::::  callout

## Staging Area

If you think of Git as taking snapshots of changes over the life of a project,
checking `Staged` (`git add`) specifies *what* will go in a snapshot
(putting things in the staging area),
and `Commit`(`git commit`) then *actually takes* the snapshot, and
makes a permanent record of it (as a commit).
If you don't have anything staged when you try to `Commit`,
Git will prompt you to use `git commit -a` or `git commit --all`,
which is kind of like gathering *everyone* to take a group photo!
However, it's almost always better to
explicitly add things to the staging area, because you might
commit changes you forgot you made. (Going back to the group photo simile,
you might get an extra with incomplete makeup walking on
the stage for the picture because you used `-a`!)
Try to stage things manually,
or you might find yourself searching for "git undo commit" more
than you would like!

::::::::::::::::::::::::::::::::::::::::::::::::::

> ![](fig/git-staging-area.svg){alt='A diagram showing how "git add" registers changes in the staging area, while "git commit" moves changes from the staging area to the repository'}

Reviewer: ```suggestion
![](fig/git-staging-area.svg){alt='A diagram showing how `git add` registers changes in the staging area, while `git commit` moves changes from the staging area to the repository'}
```
I'm not sure whether this will work inside an alt message, but I think it would be helpful for the image description to mark it as commands.

Let's watch as our changes to a file move from our editor
to the staging area and into long-term storage.

> First, let's modify our data files list by changing 'dig_health_hub_amr.csv' to 'dig_health_hub_amr_v2.csv':

Reviewer: ```suggestion
First, let's modify our data files list by changing `dig_health_hub_amr.csv` to `dig_health_hub_amr_v2.csv`:
```
So file names are also shown as monospace.

If we select `Diff index.qmd` our change will be highlighted:

![](fig/rs_git_line_chnage.svg){alt='A screenshot of RStudio highlighting changes to index.qmd'}

So far, so good:
we've replaced one line (shown with a `-` in the first column) with a new line
(shown with a `+` in the first column).

> We also have a new options: `Stage chunk` and `Discard chunk`.

Reviewer: if those are buttons, then maybe shown as: 
```suggestion
We also have a new options: <kbd>Stage chunk</kbd> and <kbd>Discard chunk</kbd>.
```
may be nicer.
